src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -240,23 +298,182 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
+    nonstd::optional<IndexMap> existingVocabulary, 
+    nonstd::optional<std::uint32_t> topKTerms, 
+    std::float_t minDf, 
+    std::float_t maxDf
+) :
+    BaseType(
+        std::move(pAllColumnAnnotations),
+        std::move(colIndex),
+        true,
+        std::move(decorator),
+        std::move(existingVocabulary),
+        std::move(topKTerms),
+        std::move(minDf),
+        std::move(maxDf)
+    ) {
 }
 
-inline Details::DocumentStatisticsTrainingOnlyPolicy::DocumentStatisticsTrainingOnlyPolicy(StringDecorator decorator) :
+// ----------------------------------------------------------------------
+// |
+// |  Anonymous namespace
+// |
+// ----------------------------------------------------------------------
+namespace {
+
+inline Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap PruneTermFreqMap(Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap & termFrequency, 
+                                                                                    std::float_t minDf, 
+                                                                                    std::float_t maxDf, 
+                                                                                    std::float_t totalNumDocumentsFloat,
+                                                                                    nonstd::optional<std::uint32_t> topKTerms) {
+
+    using FrequencyMap = Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap;
+    
+    //trim by minDf and maxDf
+    if (minDf > 0.0f || maxDf < 1.0f) {
+        for (FrequencyMap::iterator termFrequencyIter = termFrequency.begin(); termFrequencyIter != termFrequency.end();) {
+            std::float_t const freq = termFrequencyIter->second / totalNumDocumentsFloat;
+
+            if (freq < minDf || freq > maxDf) {
+                termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+            }       
+        }
+    }
+
+    //trim by topKTerms
+    if (!topKTerms.has_value() || termFrequency.empty())
+        return termFrequency;
+
+    std::uint32_t const & topKTermsValue(*topKTerms);
+
+    if (topKTermsValue > termFrequency.size())
+        throw std::invalid_argument("topKTermsUpperBound");
+
+    std::vector<std::uint32_t> frequencyVector;
+    frequencyVector.reserve(termFrequency.size());
+
+    std::transform(termFrequency.begin(), 
+                   termFrequency.end(), 
+                   std::back_inserter(frequencyVector), 
+                   [](auto const & kv) { 
+                       return kv.second; 
+                   });
+
+    //partial sort, ave O(n)
+    std::nth_element(frequencyVector.begin(), 
+                     frequencyVector.begin() + static_cast<int>(topKTermsValue) - 1, 
+                     frequencyVector.end(), 
+                     std::greater<std::uint32_t>());
+
+    std::uint32_t targetValue = frequencyVector[topKTermsValue - 1];
```

Comment:
I believe that the frequencyVector is an implementation detail used to calculate the remainingTargetValue; if so, we don't need to keep frequencyVector around once that value is calculated. Consider putting this calculation in a lambda so that the vector is destroyed once that lambda is finished.

Reply:
sure